### DISCUSSION TITLE
Cross-probing (unable to map between models)

### DISCUSSION THREAD

**Discussion by RameshwarShinde** [iTwin/itwinjs-core#3806] [Q&A]
Hi,

As I have implemented Cross Probing.

**Scenario is :** 
In [the code](https://www.itwinjs.org/sandboxes/iTwinPlatform/Cross-Probing), I have replaced the existing 2d model / P&Id model in the regarding viewport with an another P&Id model (i.e an existing sample model from 'AutoCAD plant 3d' App). It's getting loaded.
I want to add mapping between their elements (i.e. between the existing 3d model & my externally added 2d model/P&Id model) but unable to get mapping between them.

**Question is :**
Is it possible to map elements from such different model as I'm trying? If yes (maybe by manually matching elements), then how can I achieve that?

Please help & let me know If I'm trying for any wrong direction. Thanks in advance.

**Comment by pmconne** [comment]
Look in CrossProbingApi.ts. The sample uses information (relationships) stored in the iModel to populate its mappings from 2d to 3d elements. If the relationships between your own elements are not stored in the iModel, then you will have to populate your mappings from some other source. Do you have another source of such information?

**Reply by RameshwarShinde** [reply]
Okay let me check. 
What you mean exactly from the other source?
Can you push some more details plz ?

Don't have any other source of information.

**Reply by pmconne** [reply]
You are going to need to be more specific about what you need. I cannot see your data.

**Reply by RameshwarShinde** [reply]
Actually,
I have two separate models (3d & 2d/ P&Id model).
 I tried the query on 3d model `SELECT * FROM bis.ViewDefinition2d` and found results empty. So I tried getting some 2d model (P&ID model) for which I'm getting results for above query.
So now if I want to create mapping between 2d & 3d models, How I can do so?

**Reply by pmconne** [reply]
"model" or "iModel"?
An **iModel** can contain any number of 3d models and any number of 2d models.
A 3d **model** can never contain view definitions.
An iModel containing only 3d models won't contain any 2d views.
Explain what "I tried getting some 2d model" means. You tried (did you succeed?) to import (using what process?) a 2d model (from where?) into your iModel?

The relationships between 2d and 3d elements are generally produced by a connector which imports both the 2d and 3d models at the same time.

**Reply by RameshwarShinde** [reply]
'I tried getting some 2d models' means --

I have used an sample model present in 'AutoCAD plant 3d' App. I have successfully loaded that model in the viewport ( its not loaded in any iModel. It's a separate P&Id model. These two models I'm referring are separate)

And can you plz elaborate more about that connecter you are suggesting? Will that be used to produce the relationship between two separate 2d & 3d model (not from the same iModel)?

**Reply by donhfu** [reply]
Can you clarify how you obtain the sample model from AutoCAD Plant3d?  Specifically, is it a DWG file or is it an iModel that is created by a iModel Connector from your source Plant3d DWG file?

**Reply by donhfu** [reply]
You can try "iTwin Synchronizer Client" to create an iModel.  Here is an article to get you started:
https://communities.bentley.com/products/digital-twin-cloud-services/itwin-services/w/synchronization-wiki/47597/itwin-synchronizer-client
You can select AutoCAD connector to connect your DWG file to iTwin.  It is not a Plant3d connector so Plant3d business data will not be created, but Plant3d entities will be present in iModel.  The iModel shall have a bis.ViewDefinition3d from your modelspace view.  Your PID entities should be bis.PhysicalObject's on XY-plane, much like they are in Plant3d.

**Reply by RameshwarShinde** [reply]
> Can you clarify how you obtain the sample model from AutoCAD Plant3d? Specifically, is it a DWG file or is it an iModel that is created by a iModel Connector from your source Plant3d DWG file?

Okay, To be more clear,
I have not created/designed any model in 'AutoCAD Plant 3D'. Some samples are always present there when you install Plant 3D. I have used one of those directly. It was a DWG file. To use it, I have uploaded it to iTwin platform via [iTwin Synchronizer](https://www.bentley.com/en/resources/itwin-synchronizer).  From there using as  iModel.

**Reply by donhfu** [reply]
Now that you have an iModel to work with you can write your code to do whatever with it.  Your sample DWG file from Plant3d is always in 3D therefore by default you will get a 3D modelin iModel.  If you prefer, you can config AutoCAD connector in iTwin Synchronizer to treat the 3D model as 2D if the DWG extents have 0 on Z-coordinate.  If you keep the default 3D model, now you should be able to query bis.ViewDefinition3d.

**Reply by RameshwarShinde** [reply]
Yes, I'm able to query bis.ViewDefinition3d on the model (P&ID model). Getting results (containing 2 rows). But yet unable to map it with the 3d model.

**Reply by roopksaini** [reply]
Hey @RameshwarShinde, here's a [blog post ](https://medium.com/itwinjs/hablas-bis-90e6f99c8ac2) that goes into discovering the 2D / 3D relationship in an iModel sourced from OpenPlant. Now your dataset is likely different, but discovering relationships like these can require some data prodding. It could be something as simple as matching element labels (of the 2D/3D counterparts)...to concrete relationships explicitly defined in the iModel schema (as in the blog post). I suggest testing your dataset using the [iModel console](https://imodelconsole.bentley.com/) and schema explorer to discover the relationships you need. This process has been covered in the post.

**Reply by RameshwarShinde** [reply]
All right. Let me pick it up.
